pkg/cluster/registerWithHive.go

```diff
@@ -0,0 +1,108 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/openshift/installer/pkg/asset/installconfig/azure"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/hive"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+)
+
+func (m *manager) collectDataForHive(ctx context.Context) (*hive.WorkloadCluster, error) {
+	m.log.Info("collecting registration data from the new cluster")
+	if m.hiveClusterManager == nil {
+		return nil, errors.New("no hive cluster manager, skipping")
+	}
+
+	cv, err := m.configcli.ConfigV1().ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	clusterID := cv.Spec.ClusterID
+
+	// TODO: When hive support first party principles we'll need to send both first party and cluster service principles
+	servicePrincipal := azure.Credentials{
+		TenantID:       m.subscriptionDoc.Subscription.Properties.TenantID,
+		SubscriptionID: m.subscriptionDoc.ID,
+		ClientID:       m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile.ClientID,
+		ClientSecret:   string(m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile.ClientSecret),
+	}
+
+	spBytes, err := json.Marshal(servicePrincipal)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeconfig := string(m.doc.OpenShiftCluster.Properties.AROServiceKubeconfig)
+
+	resourceGroupName := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+
+	hiveWorkloadCluster := &hive.WorkloadCluster{
+		SubscriptionID:    m.subscriptionDoc.ID,
+		ClusterName:       m.doc.OpenShiftCluster.Name,
+		ResourceGroupName: resourceGroupName,
+		Location:          m.doc.OpenShiftCluster.Location,
+		InfraID:           m.doc.OpenShiftCluster.Properties.InfraID,
+		ClusterID:         string(clusterID),
+		KubeConfig:        kubeconfig,
+		ServicePrincipal:  string(spBytes),
+	}
+
+	return hiveWorkloadCluster, nil
+}
+
+func (m *manager) registerWithHive(ctx context.Context, workloadCluster *hive.WorkloadCluster) error {
+	m.log.Info("registering with hive")
+	if m.hiveClusterManager == nil {
+		return errors.New("no hive cluster manager, skipping")
+	}
+
+	cd, err := m.hiveClusterManager.Register(ctx, workloadCluster)
+	if err != nil {
+		return err
+	}
+
+	m.doc, err = m.db.PatchWithLease(ctx, m.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
```

Comment:
we can get rid of this by using the doc UUID

Reply:
Resolving thread in favour of https://github.com/Azure/ARO-RP/pull/2145#discussion_r913319980